Srcs/heiftojpeg/heiftojpeg.cpp

```diff
@@ -0,0 +1,183 @@
+        } else if (property.type == ImageFileReaderInterface::ItemPropertyType::OINF) {
+            cout << "OINF ";
+            // cout << reader.getPropertyOinf(contextId, property.index);
+        } else if (property.type == ImageFileReaderInterface::ItemPropertyType::RLOC) {
+            cout << "RLOC ";
+            cout << reader.getPropertyRloc(contextId, property.index).horizontalOffset;
+        } else if (property.type == ImageFileReaderInterface::ItemPropertyType::TOLS) {
+            cout << "TOLS ";
+            cout << reader.getPropertyTols(contextId, property.index).targetOlsIndex;
+        } else {
+            cout << "unknown ";
+        }
+        cout << "\n";
+    }
+
+    // Always reuse the parameter set from the first tile, sometimes tile 7 or 8 is corrupted
+    HevcImageFileReader::ParameterSetMap parameterSet;
+    reader.getDecoderParameterSets(contextId, tileItemIds.at(0), parameterSet);
+    std::string codeType = reader.getDecoderCodeType(contextId, tileItemIds.at(0));
+    ImageFileReaderInterface::DataVector parametersData;
+    if ((codeType == "hvc1") || (codeType == "lhv1")) {
+        // VPS (HEVC specific)
+        parametersData.insert(parametersData.end(), parameterSet.at("VPS").begin(), parameterSet.at("VPS").end());
+    }
+
+    if ((codeType == "avc1") || (codeType == "hvc1") || (codeType == "lhv1")) {
+        // SPS and PPS
+        parametersData.insert(parametersData.end(), parameterSet.at("SPS").begin(), parameterSet.at("SPS").end());
+        parametersData.insert(parametersData.end(), parameterSet.at("PPS").begin(), parameterSet.at("PPS").end());
+    } else {
+        // No other code types supported
+        // throw FileReaderException(FileReaderException::StatusCode::UNSUPPORTED_CODE_TYPE);
+    }
+
+    ImageFileReaderInterface::DataVector itemDataWithDecoderParameters;
+    ImageFileReaderInterface::DataVector itemData;
+
+    std::vector<Magick::Image> tileImages;
+    for (auto& tileItemId: tileItemIds) {
+        itemDataWithDecoderParameters.clear();
+        itemDataWithDecoderParameters.insert(itemDataWithDecoderParameters.end(), parametersData.begin(), parametersData.end());
+
+        itemData.clear();
+        reader.getItemData(contextId, tileItemId, itemData);
+
+        // +1 comes from skipping first zero after decoder parameters
+        itemDataWithDecoderParameters.insert(itemDataWithDecoderParameters.end(), itemData.begin() + 1, itemData.end());
+
+        Magick::Image image;
+        decodeData(itemDataWithDecoderParameters, &image);
+        tileImages.push_back(image);
+    }
+
+    Magick::Montage montageOptions;
+    montageOptions.tile("8x6");
+    montageOptions.geometry("512x512");
+    std::list<Magick::Image> montage;
+    Magick::montageImages(&montage, tileImages.begin(), tileImages.end(), montageOptions);
+    Magick::Image image = montage.front();
+    image.magick("JPEG");
+    image.crop(Magick::Geometry(gridItem.outputWidth,gridItem.outputHeight));
```

Comment:
is outputWidth/Height the total original image size or just for that grid item?  we should probably scale down to 2048x1536 while we're at it here

Reply:
that's the composed size of the image. in this case the grid is in 6x8 blocks of 512x512 each, but the actual image is slightly smaller than 4096x3072 (I'm not sure why). The outputWidth and outputHeight indicate the correct dimensions of 4032x3024, so we crop to that to avoid a black bar along the right and bottom of the image.

Reply:
Does this correctly handle photos taken as a square, or post-cropped to something other than 4:3?

Reply:
I added a command-line option `-s <max_size>` which allows you to specify the maximum pixel dimension. The image is scaled with aspect preservation to fit that dimension. So we can invoke the program with `-s 2048` to get your desired scale.

Reply:
@bcherry you should test with "photos taken as a square, or post-cropped to something other than 4:3" to make sure they end up cropped correctly given the black bar bug on the default aspect ratio